File: session13/GraphList.java
import java.util.*;

public class GraphList {
	private static class Edge {
    int v2;
		double w;
		public Edge(int v2, double w) { this.v2 = v2; this.w = w; }
	}

	private ArrayList< LinkedList<Edge> > edges;
	private int V;

	public GraphList(int V) {
		this.V = V;
		edges = new ArrayList<>(V);
		for (int i = 0; i < V; i++)
			edges.add(new LinkedList<Edge>());
	}
	public void set(int i, int j, double w) {
		LinkedList<Edge> edgei = edges.get(i);
		for (Edge e : edgei) {
			if (e.v2 == j) {
				e.w = w;
				return;
			}
		}
		edgei.add(new Edge(j, w));
	}
	public double get(int i, int j) {
		for (Edge e : edges.get(i)) {
			if (e.v2 == j)
				return e.w;
		}
		return Double.POSITIVE_INFINITY;
	}
	public void DFS(int v) {
		boolean[] visited = new boolean[V];
		DFS2(v, visited);
	}
	private void DFS2(int v, boolean[] visited) {
    visited[v] = true;
		System.out.print(v);
		for (Edge e : edges.get(v) ) {
			if (!visited[e.v2]) {
				DFS2(e.v2, visited);
			}
		}
	}
	public void iterativeDFS(int v) {
		boolean[] visited = new boolean[V];
		Stack<Integer> todo = new Stack<>();
    todo.push(v);
		visited[v] = true;;
		while (!todo.isEmpty()) {
			int v2 = todo.pop();
			System.out.print(v2);
			LinkedList<Edge> adjacent = edges.get(v2);
			for (Edge e : adjacent) {
				if (!visited[e.v2]) {
					todo.push(e.v2);
					visited[e.v2] = true;
				}
			}
		}
	}

	public static void main(String[] args) {
    GraphList gl = new GraphList(5);
		gl.set(0, 1, 1.5);
		gl.set(1, 2, 2.0);
		gl.set(2, 3, 2.5);
		gl.set(3, 4, 3.0);
		gl.DFS(0);
		System.out.println();
		gl.iterativeDFS(0);
		System.out.println();
	}
}

File: old/session13/GraphList.cc
#include <iostream>
#include <stack>
using namespace std;

class GraphList {
private:
  //in Java: Double.POSITIVE_INFINITY
    public static constexpr double INFINITY = 1.0 / 0.0;
  class Edge {
  public:
    int to;
    double w;
    Edge* next;
    Edge(int to, double w, Edge* next) : to(to), w(w), next(next) {}
  };
  Edge** edges;
  int V;
public:
  GraphList(int V) : V(V) {
    edges = new Edge*[V];
  }
  ~GraphList() {
    Edge** p = edges;
    for (int i = V; i > 0; i--, ++p) {
      // delete the linked list...
      Edge* temp;
      for (Edge* q = *p; q != nullptr; q = temp) {
	temp = q->next;
	delete q;
      }
    }
    delete [] edges;
  }
  void setEdge(int i, int j, double w) {
    for (Edge* e = edges[i]; e != nullptr; e = e->next) {
      if (e->to == j) {
	e->w = w;
	return;
      }
    }
    edges[i] = new Edge(i,j, w, edges[i]); // addFirst
  }
  
  double adjacent(int  i, int j) { //O(V)
    Edge* n = edges[i];
    if (n == nullptr)
      return INFINITY;
    for ( ; n != nullptr; n=n->next)
      if (n->to == j)
	return n->w;
    return INFINITY;
  }
  void DFS(int v) {
    bool visited[V];
    for (int i = 0; i < V; i++)
      visited[V] = false;
    stack<int> todo;
    visited[v] = true; 
    todo.push(v);
    while (!todo.empty()) {
      int v = todo.back();
      todo.pop_back();
      cout << v;
      for (Edge* p = edges[v]; p != nullptr; p = p->next) {
	int v2 = p->to;
	if (!visited[v2]) {
	  stack.push_back(v2);
	  visited[v2] = true;
	}
      }
    }
  }
};
